src/common/lua/ngx_wasm_lua_ffi.c

```diff
@@ -304,126 +288,126 @@ ngx_wa_ffi_shm_setup_zones(ngx_wa_ffi_shm_setup_zones_handler_pt setup_handler)
+retrieve_keys(ngx_rbtree_node_t *node, ngx_rbtree_node_t *sentinel,
+    ngx_uint_t limit, ngx_uint_t start_idx, ngx_uint_t *cur_idx,
+    ngx_str_t **keys)
 {
-    ngx_wa_shm_kv_node_t *n = (ngx_wa_shm_kv_node_t *) node;
+    ngx_wa_shm_kv_node_t  *n = (ngx_wa_shm_kv_node_t *) node;
 
-    if (limit > 0 && (*total - start) == limit) {
+    if ((*cur_idx - start_idx) == limit) {
         return;
     }
 
-    if (keys && *total >= start) {
-        keys[(*total - start)] = &n->key.str;
+    if (*cur_idx >= start_idx) {
+        /* append to result */
+        keys[(*cur_idx - start_idx)] = &n->key.str;
     }
 
-    (*total)++;
+    (*cur_idx)++;
 
     if (node->left != sentinel) {
-        shm_kv_retrieve_keys(node->left, sentinel, start, limit, keys, total);
+        retrieve_keys(node->left, sentinel, limit, start_idx, cur_idx, keys);
     }
 
     if (node->right != sentinel) {
-        shm_kv_retrieve_keys(node->right, sentinel, start, limit, keys, total);
+        retrieve_keys(node->right, sentinel, limit, start_idx, cur_idx, keys);
     }
 }
 
 
 ngx_int_t
-ngx_wa_ffi_shm_iterate_keys(ngx_wa_shm_t *shm,
-                            ngx_uint_t page, ngx_uint_t page_size,
-                            ngx_str_t **keys, ngx_uint_t *total)
+ngx_wa_ffi_shm_iterate_keys(ngx_wa_shm_t *shm, ngx_uint_t page_size,
+    ngx_uint_t *clast_idx, ngx_uint_t *cur_idx, ngx_str_t **keys)
 {
+    ngx_uint_t        last_idx = clast_idx ? *clast_idx : 0;
     ngx_wa_shm_kv_t  *kv;
 
-    ngx_wa_assert(shm && shm->type != NGX_WA_SHM_TYPE_QUEUE);
+    ngx_wa_assert(shm->type != NGX_WA_SHM_TYPE_QUEUE);
 
     if (!ngx_wa_shm_locked(shm)) {
         return NGX_ABORT;
     }
 
     kv = shm->data;
 
-    if (page >= kv->nelts) {
-        return NGX_DECLINED;
+    if (last_idx >= kv->nelts) {
+        /* finished */
+        return NGX_DONE;
     }
 
-    if (kv->rbtree.root != kv->rbtree.sentinel) {
-        shm_kv_retrieve_keys(kv->rbtree.root, kv->rbtree.sentinel,
-                             page, page_size, keys, total);
-    }
+    /* cannot be empty here (kv->nelts > 0) */
+    ngx_wa_assert(kv->rbtree.root != kv->rbtree.sentinel);
 
-    *total = *total - page;
+    retrieve_keys(kv->rbtree.root, kv->rbtree.sentinel, page_size, last_idx,
+                  cur_idx, keys);
+
+    *cur_idx -= last_idx;
+
+    if (clast_idx) {
+        *clast_idx += *cur_idx;
+    }
 
     return NGX_OK;
 }
 
 
+ngx_uint_t
+ngx_wa_ffi_shm_kv_nelts(ngx_wa_shm_t *shm)
+{
+    ngx_wa_shm_kv_t  *kv;
+
+    ngx_wa_assert(shm->type != NGX_WA_SHM_TYPE_QUEUE);
+
+    kv = shm->data;
+
+    return kv->nelts;
+}
+
+
 ngx_int_t
-ngx_wa_ffi_shm_get_kv_value(ngx_wa_shm_t *shm,
-                            ngx_str_t *k, ngx_str_t **v, uint32_t *cas)
+ngx_wa_ffi_shm_kv_get(ngx_wa_shm_t *shm, ngx_str_t *k, ngx_str_t **v,
+    uint32_t *cas)
 {
-    unsigned   unlock = 0;
     ngx_int_t  rc;
 
-    ngx_wa_assert(shm && k && v && cas);
+    ngx_wa_assert(shm->type == NGX_WA_SHM_TYPE_KV);
```

Comment:
It's not clear to me the criteria we should use to decide if something should be asserted.

I thought we had agreed to only check for pointers validity.
The current assertion seems unnecessary as `ngx_wa_ffi_shm_kv_get` is bound to only be called with `NGX_WA_SHM_TYPE_KV` shms from `shm.lua`. That being said, `shm.lua` also guarantees that `ngx_wa_ffi_shm_get_kv_value` will only receive valid pointers.

I also don't think these assertions add any substantial amount of documentation to what these functions do.

Reply:
If a pointer is invalid, the function will crash later and it will be obvious in the produced coredump. Having assertions that contain several conditions within them also require the coredump to be opened in the first place (since the code line won't be enough to determine the failing condition). It can also be argued that according to the code in `shm.lua`, `k`, `v,` and `cas` would never be invalid when reaching this code path, so as per that argument the original assertion was unnecessary in the first place as well. As it stands today it isn't really necessary either but it makes the function more explicit to the reader so I kept it from another function that checked the type (assertions were inconsistent) and consistently applied it to all functions.

Reply:
No strong feelings on this from my end, we can go back to super-explicit assertions too.

Reply:
It hasn't occurred to me that having multiple conditions in a single assert makes it harder to distinguish which one was false. Good point. 

Thank you for the thoughtful explanation.